operator/pkg/functional/map.go

```diff
@@ -16,3 +16,69 @@ func MapFn[T any, U any](fn func(T) U, a []T) []U {
 	}
 	return s
 }
+
+func deepCopyMap(m map[string]any) map[string]any {
+	copied := make(map[string]any)
+	for k, v := range m {
+		switch cast := v.(type) {
+		case map[string]any:
+			copied[k] = deepCopyMap(cast)
+		case []any:
+			copied[k] = deepCopyElements(cast)
+		default:
+			copied[k] = v
+		}
+	}
+	return copied
+}
+
+func deepCopy(v any) any {
+	switch cast := v.(type) {
+	case map[string]any:
+		return deepCopyMap(cast)
+	case []any:
+		return deepCopyElements(cast)
+	default:
+		return v
+	}
+}
+
+func deepCopyElements(v []any) []any {
+	copied := make([]any, len(v))
+	for _, value := range v {
+		switch cast := value.(type) {
+		case map[string]any:
+			copied = append(copied, deepCopyMap(cast))
+		case []any:
+			copied = append(copied, deepCopyElements(cast))
+		default:
+			return v
+		}
+	}
+
+	return copied
+}
+
+func MergeMaps(first, second map[string]any) map[string]any {
+	merged := deepCopyMap(first)
+	for k, v := range second {
+		if original, ok := merged[k]; ok {
+			switch cast := original.(type) {
+			case map[string]any:
+				// the types must match, otherwise we can't merge them
+				if vmap, ok := v.(map[string]any); ok {
+					merged[k] = MergeMaps(cast, deepCopyMap(vmap))
+				}
+			case []any:
+				// the types must match, otherwise we can't merge them
+				if varray, ok := v.([]any); ok {
+					merged[k] = append(cast, deepCopyElements(varray))
```

Comment:
If this is meant to match the behavior of helm's value merging they don't merge arrays AFAIK. I'd strongly recommend delegating to their internals were possible or using `rapid` to assert that the behavior between our implementation and theirs is the same to prevent accidental divergences.

Reply:
So, currently I'm just using this in testing. I did originally use it for merging materialized `Helm` values, but realized I can actually just delegate all of that to the Redpanda --> PartialValues normalization code in the helm chart, so I can get away with just pulling the equivalent of `values.yaml` for doing stuff via the `Helm` artifacts.

I'm ok if we don't want to introduce something like this here just refactoring the tests to use something simpler too.

Reply:
Up to you! I didn't realize this was just for testing. I'd lean slightly more towards using the PartialValues normalization code so we don't accidentally pick this up in a helm code path but I'm also okay with a lot of red tape in the comments.